cfg <- list(

	etafile = "etaValues.bmat",
		#The file containing the relative eta values for each codon

	genomefile = "ecoli.fasta",
		#file containing the genome in question

	optimalfile = "optimalecoli.fasta",
	pessimalfile = "pessimalecoli.fasta"
		#files where the optimal and/or pessimal genomes are written
)
